lib/protobuf/field/field_array.rb

```diff
@@ -21,21 +21,51 @@ def initialize(field)
       #
 
       def []=(nth, val)
-        super(nth, normalize(val)) unless val.nil?
+        if nth < 0 || nth > size
+          fail IndexError, "Index #{nth} is outside valid range 0...#{size}"
+        end
+        super(nth, normalize(val))
       end
 
       def <<(val)
-        super(normalize(val)) unless val.nil?
+        super(normalize(val))
       end
 
-      def push(val)
-        super(normalize(val)) unless val.nil?
+      alias_method :push, :<<
+
+      def unshift(val)
+        super(normalize(val))
       end
 
       def replace(val)
         raise_type_error(val) unless val.is_a?(Array)
-        val.map! { |v| normalize(v) }
-        super(val)
+        super(val.map { |v| normalize(v) })
+      end
+
+      def map!(&block)
+        if block_given?
+          new_ary = map(&block)
+          replace(new_ary)
+        else
+          super()
+        end
+      end
+
+      def fill(*args, &block)
```

Comment:
OK just tested this against your changes -- `fill` seems to throw a `TypeError` rather than an `IndexError` and `insert` just lets it happen